File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.py
import sys
input = sys.stdin.readline


n = int(input())

triangle = []

for _ in range(n):
  triangle.append(list(map(int, input().split())))

for i in range(1,n):
  for j in range(i+1):

    # 맨 왼쪽의 수
    if j==0:
      triangle[i][j] += triangle[i-1][j]

    # 맨 오른쪽의 수
    elif i == j:
      triangle[i][j] += triangle[i-1][j-1]

    # 둘 다 아니면 대각선 왼쪽과 대각선 오른쪽의 값을 더한 것 중 큰 수로
    else:
      triangle[i][j] += max(triangle[i - 1][j - 1], triangle[i - 1][j])
 
print(max(triangle[n-1]))

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {

    for ( let i = 1 ; i < triangle.length ; i++) {
        for ( let j = 0 ; j < triangle[i].length ; j++) {
            // 삼각형의 왼쪽 또는 오른쪽 대각선에 있는 숫자면
            // 이동 가능한 칸이 한 칸밖에 없으니까 그대로 더해주기
            if ( j === 0 ) {
                triangle[i][j] += triangle[i-1][j];
            }
            else if ( j === triangle[i].length - 1 ) {
                triangle[i][j] += triangle[i-1][j-1];
            // 대각선이 아닌 내부의 숫자들은 위에서 더 큰 수를 비교해서 더해주기
            }
            else {
                triangle[i][j] += Math.max(triangle[i-1][j-1], triangle[i-1][j]);
            }
        }
    }

    var answer = Math.max(...triangle[triangle.length-1]);

    return answer;
}